File: app/src/main/java/com/rubyhuntersky/peregrine/model/FundingAccount.java
package com.rubyhuntersky.peregrine.model;

import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.math.BigDecimal;
import java.util.List;

/**
 * @author wehjin
 * @since 2/6/16.
 */
public interface FundingAccount extends Parcelable {
    @NonNull
    String getAccountName();
    @NonNull
    BigDecimal getCashAvailable();
    @NonNull
    List<FundingOption> getFundingOptions(String exclude);

    boolean hasFundsForBuyIntention(@NonNull BuyIntention buyIntention);

    @NonNull
    FundingStatus getStatusForBuy(@NonNull BuyIntention buyIntention);
}

File: app/src/main/java/com/rubyhuntersky/peregrine/model/FundingAccount.kt
package com.rubyhuntersky.peregrine.model

import android.os.Parcelable
import com.rubyhuntersky.peregrine.utility.toIntString
import java.math.BigDecimal

interface FundingAccount : Parcelable {
    val accountName: String
    val cashAvailable: BigDecimal
    fun getFundingOptions(exclude: String): List<FundingOption>
    fun hasFundsForBuyIntention(buyIntention: BuyIntention): Boolean
    fun getStatusForBuy(buyIntention: BuyIntention): FundingStatus

    fun hasFundsForPurchase(purchaseAmount: BigDecimal): Boolean {
        return cashAvailable >= purchaseAmount
    }

    fun getSharesPurchasableWithCash(product: AssetNamePrice): String {
        return cashAvailable.divide(product.price, Values.SCALE, BigDecimal.ROUND_HALF_UP).toIntString()
    }

    fun isFullyFundedForPurchase(purchaseAmount: BigDecimal): Boolean {
        return getShortfall(purchaseAmount) <= BigDecimal.ZERO
    }

    fun getShortfall(purchaseAmount: BigDecimal): BigDecimal {
        return purchaseAmount.subtract(cashAvailable)
    }
}
